src/search/Encoding.ts

```diff
@@ -18,38 +18,130 @@
 
 import { ExecutionResult } from "./ExecutionResult";
 import { ObjectiveFunction } from "./objective/ObjectiveFunction";
+import {prng} from "../util/prng";
+import {getUserInterface} from "../ui/UserInterface";
+import {EncodingSampler} from "./EncodingSampler";
+import {Decoder} from "./Decoder";
 
 /**
  * Encoding of the search problem.
  *
  * @author Mitchell Olsthoorn
  */
-export interface Encoding {
+export abstract class Encoding {
+  protected _crowdingDistance: number;
+  protected _rank: number;
+  protected _id: string;
+  protected _assertions: Map<string, string>;
+
   /**
-   * Store the distance to an objective for this encoding.
-   *
-   * @param objectiveFunction The objective
-   * @param distance The distance
+   * Mapping from objective to their distance values for this test case.
+   * @protected
    */
-  setDistance(
-    objectiveFunction: ObjectiveFunction<Encoding>,
-    distance: number
-  ): void;
+  protected _objectives: Map<ObjectiveFunction<Encoding>, number>;
+
+  /**
+   * The last execution result of this test case.
+   * @protected
+   */
+  protected _executionResult: ExecutionResult;
+
+  /**
+   * Constructor.
+   */
+  constructor() {
+    this._crowdingDistance = 0;
+    this._rank = 0;
+    this._id = prng.uniqueId(20);
+    this._objectives = new Map<ObjectiveFunction<Encoding>, number>();
+    this._assertions = new Map();
+    getUserInterface().debug(`Created test case: ${this._id}`);
+  }
+
+
+  abstract mutate(sampler: EncodingSampler<Encoding>): Encoding;
+
+  abstract hashCode(decoder: Decoder<Encoding, string>): number;
+
+  getCrowdingDistance(): number {
+    return this._crowdingDistance;
+  }
+
+  setCrowdingDistance(value: number): void {
+    this._crowdingDistance = value;
+  }
+
+  getRank(): number {
+    return this._rank;
+  }
+
+  setRank(value: number): void {
+    this._rank = value;
+  }
+
+  get id(): string {
+    return this._id;
+  }
+
+  get assertions(): Map<string, string> {
+    return this._assertions;
+  }
+
+  set assertions(value: Map<string, string>) {
```

Comment:
Is not immutable

Reply:
Currently it is not possible to change this
For now, the Encoding does have mutable parts.
The genes of the encoding are immutable